### PR TITLE
[no ticket][risk=no] Puppeteer modify each test to use same workspace

### PR DESCRIPTION
**Pull request by aweng98**
Background:
For most e2e tests, a new workspace is to be created in every test per run. A very small number of tests look for an existing workspace (older than 30 min) to use. Everyday, a Jenkins cron job deletes all workspaces created by tests from previous day.

Goal:
Reduce tests total runtime and create lesser number of workspaces.

How:
If every test can reuse an existing workspace, then we won't have to create many new workspaces.
In a test, a unique workspace name is hardcoded. If this workspace doesn't exist, then it will be created. Next time same test will find this workspace and use it.
Next day, Jenkins cron job will delete all workspaces and every test will re-create new workspace with same name.

Changes:
Updated tests code; removed delete-workspace call in each test except `workspace-create.spec.ts`.

### CODE REVIEW

File: e2e/tests/notebook/owner-copy-to-workspace.spec.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+ * Test:
+ * - Create new Workspace as the copy-to destination Workspace.
+ * - Create new Workspace as copy-from Workspace and create new notebook in this Workspace.
+ * - Run code to print WORKSPACE_NAMESPACE. It should match Workspace namespace from Workspace URL.
+ * - Copy notebook to destination Workspace and give copied notebook a new name.
+ * - Verify copied notebook is in destination Workspace.
+ * - Open copied notebook and run code to print WORKSPACE_NAMESPACE. It should match destination Workspace namespace.
+ * - Delete notebooks.
+ */
+describe('Copy notebook', () => {
```
Comment (aweng98) [inline e2e/tests/notebook/owner-copy-to-workspace.spec.ts:22]: Split from `nightly/owner-copy-to-workspace.spec.ts`.

File: e2e/tests/workspace/workspace-copy-notebook.spec.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+ * - Create new Workspace as the copy-to destination Workspace.
+ * - Create new Workspace as copy-from Workspace and create new notebook in this Workspace.
+ * - Run code to print WORKSPACE_NAMESPACE. It should match Workspace namespace from Workspace URL.
+ * - Copy notebook to destination Workspace and give copied notebook a new name.
+ * - Verify copied notebook is in destination Workspace.
+ * - Open copied notebook and run code to print WORKSPACE_NAMESPACE. It should match destination Workspace namespace.
+ * - Delete notebooks.
+ *
+ */
+describe('Copy notebook to workspace with different CDR version', () => {
```
Comment (aweng98) [inline e2e/tests/workspace/workspace-copy-notebook.spec.ts:23]: Split from `nightly/owner-copy-to-workspace.spec.ts`

File: e2e/tests/conceptsets/conceptset-edit-rename.spec.ts [hunk 3]
```diff
@@ -21,50 +22,27 @@ describe('Editing and rename Concept Set', () => {
...
-    // Verify Participant Count are numberical values
-    expect(Number.isNaN(parseInt(row1.rollUpCount.replace(/,/g, ''), 10))).toBe(false);
-    expect(Number.isNaN(parseInt(row2.rollUpCount.replace(/,/g, ''), 10))).toBe(false);
-
-    await conceptSearchPage.reviewAndSaveConceptSet();
-
-    // Save new Concept Set.
-    const conceptSetName = await conceptSearchPage.saveConceptSet(SaveOption.CreateNewSet);
+    // Create new Concept Set
+    const conceptSetName = await dataPage.createDefaultProcedures('Radiologic examination');
```
Comment (aweng98) [inline e2e/tests/conceptsets/conceptset-edit-rename.spec.ts:31]: Duplicated code here and in another test so I created new func for reuse.

File: e2e/tests/runtime/pause-resume.spec.ts [hunk 4]
```diff
@@ -17,15 +17,12 @@ describe('Updating runtime status', () => {
 
     // Create runtime
     await runtimePanel.createRuntime();
-    await page.waitForTimeout(2000);
```
Comment (aweng98) [inline e2e/tests/runtime/pause-resume.spec.ts:20]: sleep is bad.

File: e2e/app/element/base-element.ts [hunk 5]
```diff
@@ -31,15 +31,17 @@ export default class BaseElement extends Container {
...
-      logger.error(`waitForXpath('${this.xpath}') failed`);
-      logger.error(err);
-      // Debugging pause
-      // await jestPuppeteer.debug();
-      throw new Error(err);
-    }
+    return this.page
+      .waitForXPath(this.xpath, waitOptions)
+      .then((elemt) => (this.element = elemt.asElement()))
+      .catch((err) => {
```
Comment (aweng98) [inline e2e/app/element/base-element.ts:37]: use promise `catch` in favor of `try/catch`

File: e2e/tests/workspace/workspace-create.spec.ts [hunk 6]
```diff
@@ -84,15 +86,4 @@ describe('Creating new workspaces', () => {
     const dataPage1 = new WorkspaceDataPage(page);
     await dataPage1.verifyWorkspaceNameOnDataPage(newWorkspaceName);
   });
-
-  // // helper function to check visible workspace link on Data page
```
Comment (aweng98) [inline e2e/tests/workspace/workspace-create.spec.ts:88]: remove commented-out code

**Review by jmthibault79 — CHANGES_REQUESTED** [review]
Please submit a new PR which contains code to reuse workspaces and literally nothing else.

File: e2e/app/page/workspace-edit-page.ts [hunk 7]
```diff
@@ -16,7 +16,7 @@ import { makeWorkspaceName } from 'utils/str-utils';
 
 const faker = require('faker/locale/en_US');
 
-export const PageTitle = 'Create Workspace';
+export const PageTitle = 'Create|Duplicate Workspace';
```
Comment (aweng98) [inline e2e/app/page/workspace-edit-page.ts:19]: `waitForLoad()` didn't consider Duplicate Workspace page before. Duplicate and Build page are the same but the finish button are differently labeled.

**Comment by aweng98** [comment]
Removed unrelated changes.

File: e2e/tests/workspace/old-cdr-version-modal.spec.ts [hunk 8]
```diff
@@ -32,11 +32,11 @@ describe('OldCdrVersion Modal restrictions', () => {
     await editPage.clickCreateFinishButton(createButton);
   });
 
-  test('OWNER cannot duplicate workspace to an older CDR Version without consenting to restrictions', async () => {
-    const workspaceCard = await findOrCreateWorkspaceCard(page);
+  const workspace = 'e2eWorkspaceCDRTest';
```
Comment (jmthibault79) [inline e2e/tests/workspace/old-cdr-version-modal.spec.ts:35]: please make this more descriptive

File: e2e/utils/waits-utils.ts [hunk 9]
```diff
@@ -44,10 +44,11 @@ export async function waitForDocumentTitle(page: Page, titleSubstr: string): Pro
   try {
     const jsHandle = await page.waitForFunction(
       (t) => {
+        const regExp = new RegExp(t);
         const actualTitle = document.title;
-        return actualTitle.includes(t);
+        return actualTitle && regExp.test(actualTitle);
       },
-      { timeout: 10 * 60 * 1000 },
+      { timeout: 60 * 1000 },
```
Comment (jmthibault79) [inline e2e/utils/waits-utils.ts:51]: why reduce this timeout?

**Review by jmthibault79 — APPROVED** [review]
Looks good - a few questions.

I'm concerned about simultaneous tests interfering with each others' workspaces.  But this is worth a try.

File: e2e/tests/workspace/old-cdr-version-modal.spec.ts (hunk 8 above)
Comment (aweng98) [reply e2e/tests/workspace/old-cdr-version-modal.spec.ts:35]: Done. it's now `'e2eCloneWorkspaceCDRConsentTest'`.

File: e2e/utils/waits-utils.ts (hunk 9 above)
Comment (aweng98) [reply e2e/utils/waits-utils.ts:51]: I think a long timeout is hiding problem if a page really takes that long to load. IMO, no page should take 10 min to load. Also, long timeout makes longer to fail if fails here.

**Review by aweng98 — COMMENTED** [review]
This change is going to reduce number of tests failures by empty billing projects buffer. Thank you.

**Pull request merged**